### PR TITLE
[WFCORE-3411] CLI operation completion issues.

### PR DESCRIPTION
**Pull request by MMarus**
WFCORE: https://issues.jboss.org/browse/WFCORE-3411

I have encountered another completion issue which I described in first comment of the WFCORE-3411. Following lines are related to the new mentioned issue https://github.com/wildfly/wildfly-core/compare/master...MMarus:WFCORE-3411-CLI_operation_completion_issues?expand=1#diff-b5ac7c9723bb46958c770f161f4759c8R432 . I can move this into separate WFCORE issue and PR if needed.

### CODE REVIEW

File: cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java [hunk 1]
```diff
@@ -592,6 +601,9 @@ private int completeArgumentValueAndPropertyNames(CommandContext ctx, ParsedComm
                 final CommandLineFormat format = parsedCmd.getFormat();
                 if (format != null && format.getPropertyListEnd() != null && format.getPropertyListEnd().length() > 0) {
                     candidates.add(format.getPropertyListEnd());
+                    if (candidates.size() == 1) {
+                        result = parsedCmd.getLastChunkIndex() + chunk.length();
+                    }
```
Comment (xstefank) [inline cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: this can probably be refactored to a separate method

File: cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java [hunk 2]
```diff
@@ -428,8 +428,22 @@ private boolean canAppearNext(Property prop) {
...
+                    if(ctx.getParsedCommandLine().isLastPropertyNegated()) {
+                        for (Property prop : propList) {
+                            if(radical.equals(prop.getName())){
+                                radical = null;
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    radical = null;
```
Comment (xstefank) [inline cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: not necessary

File: cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java [hunk 3]
```diff
@@ -592,6 +601,9 @@ private int completeArgumentValueAndPropertyNames(CommandContext ctx, ParsedComm
                 final CommandLineFormat format = parsedCmd.getFormat();
                 if (format != null && format.getPropertyListEnd() != null && format.getPropertyListEnd().length() > 0) {
                     candidates.add(format.getPropertyListEnd());
+                    if (candidates.size() == 1) {
```
[the remark this answers is not in the document]
Comment (MMarus) [reply cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: Yes, You are right. I have already extracted this part of code, which repeats 4 times in the code into a separate method. Should I change this part to "result = parsedCmd.getLastChunkIndex() + chunk.length();" or just leave the method call as it is now in the https://github.com/wildfly/wildfly-core/pull/3155/commits/11911af7eea6dfd89c8cb2f843524a8c75230282 commit ?

File: cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java [hunk 4]
```diff
@@ -592,6 +601,9 @@ private int completeArgumentValueAndPropertyNames(CommandContext ctx, ParsedComm
...
+                        result = parsedCmd.getLastChunkIndex() + chunk.length();
```
[the remark this answers is not in the document]
Comment (MMarus) [reply cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: It should be ensured by line https://github.com/wildfly/wildfly-core/commit/11911af7eea6dfd89c8cb2f843524a8c75230282#diff-f91220412cef158ff6623a2dccc627a4R462 
where the lastArg is only populated in case the FullName of the arg is not null (.equals would fail with NPE if it was called on null) and is equal to the chunk and the line
https://github.com/wildfly/wildfly-core/commit/11911af7eea6dfd89c8cb2f843524a8c75230282#diff-f91220412cef158ff6623a2dccc627a4R554
where it is checked that lastArg is not null.

However, the lastArg and chunk are mutable, therefore in case there would be committed a code which assigns a null to either of them, this indirect "check" would fail.
Do You think it would be better to add a null check of the chunk in the new method recalculateResult() commited in 11911af ?

[the remark this answers is not in the document]
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: It would be nice to add a check and a comment.

File: cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: you can keep the method.

File: cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java [hunk 5]
```diff
@@ -428,8 +428,20 @@ private boolean canAppearNext(Property prop) {
                 final List<Property> propList = reqProps.asPropertyList();
                 result = new ArrayList<CommandArgument>(propList.size());
 
-                String radical = ctx.getParsedCommandLine().getLastParsedPropertyValue() == null
-                        ? ctx.getParsedCommandLine().getLastParsedPropertyName() : null;
+                String radical = null;
+                if(ctx.getParsedCommandLine().getLastParsedPropertyValue() == null) {
+                    radical = ctx.getParsedCommandLine().getLastParsedPropertyName();
+                    //Check if the property is completely specified and is negated
```
Comment (jfdenise) [inline cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: Why do you need to set the radical to null in this case? Would you have an example?

Comment (MMarus) [reply cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: The methond canAppearNext() [1] returns false when the property is already in the input line or when the name does not start with radical. That means if the operation e.g. :reload-servers has 2 possible properties "blocking" and "start-mode" then in case the "!blocking" is specified the radical would be set to "blocking" and the  canAppearNext() would return false for every one command, since the "blocking" is already  in the input and the other properties don't start with "blocking" string.

In the old code the radical was set to null for example in this case ":reload-servers(blocking=false<TAB>" because the last property blocking has the value false and getLastParsedPropertyValue() returned "false". 
However, In case  ":reload-servers(!blocking<TAB>" which has similar meaning as the first one, the radical was set to "blocking". This was caused because the getLastParsedPropertyValue() returned null which I think is right, since there is no explicit specified value like in the first case the "=false". Therefore when the radical was set to null in the first case I think we could do the same for the case when the property is negated and its name is completely specified e.g. "!blocking".

[1] https://github.com/MMarus/wildfly-core/blob/11911af7eea6dfd89c8cb2f843524a8c75230282/cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java#L378

File: cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MMarus) [reply cli/src/main/java/org/jboss/as/cli/operation/OperationRequestCompleter.java]: Thank you very much, I have added them.

File: cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java (hunk 5 above)
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: Thank-you for taking the time to detail the case. You are right, you perhaps have to take into account a corner case, let's imagine we have blocking and blocking-param options.
reload:(!blocking<TAB> ==> should propose "," and blocking-param. This is something that you should be able to reproduce in your new cli test.
BTW, when making changes to completion it is good to extend the test https://github.com/wildfly/wildfly-core/blob/master/testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/cli/CliCompletionTestCase.java

**Comment by jmesnil** [comment]
retest this please

File: cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java (hunk 5 above)
Comment (MMarus) [reply cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: I just looked into the test *PropertiesCompletionTestCase* again and the *DefaultOperationCandidatesProvider* is not tested there, since there is a *MockOperationCandidatesProvider* used instead. I think it's because the method *getProperties()* first creates and executes request in order to get list of properties which wouldn't be easy to mock. However, in order to create the resulting list of  *CommandArgument* only the *Property* list is needed (line 429-524 [1]) which could be very easily moved to separate method and it would allow to test this functionality.

Do You think I can move that part into separate method or is it too big change ?
Thank You for answer.

[1] https://github.com/MMarus/wildfly-core/blob/80f7c778602d3be0b05d0cc89d20a7efd4c89c0b/cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java#L429

Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java]: @MMarus , +1 to increase unit-testability of completion!

**Comment by MMarus** [comment]
I have moved (lines 429-524 [1]) to separate method and added the tests mentioned in the comments. 

 For the corner case,when we have blocking and blocking-param options I have created 2 separate tests to test the relevant logic in relevant classes (*PropertiesCompletionTestCase* and *DefaultOperationCandidatesProviderTestCase*).

[1] https://github.com/MMarus/wildfly-core/blob/80f7c778602d3be0b05d0cc89d20a7efd4c89c0b/cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java#L429

**Comment by MMarus** [comment]
retest this please

**Comment by jmesnil** [comment]
@jfdenise are you ok with the changes you requested?
